localstack-core/localstack/services/lambda_/provider.py

```diff
@@ -4118,87 +4141,141 @@ def delete_function_concurrency(
     # =======================================
     # ===============  TAGS   ===============
     # =======================================
-    # only function ARNs are available for tagging
+    # only Function, Event Source Mapping, and Code Signing Config (not currently supported by LocalStack) ARNs an are available for tagging in AWS
 
-    def _get_tags(self, function: Function) -> dict[str, str]:
-        return function.tags or {}
+    def _get_tags(self, resource: TaggableResource) -> dict[str, str]:
+        state = self.fetch_lambda_store_for_tagging(resource)
+        lambda_adapted_tags = {
+            tag["Key"]: tag["Value"]
+            for tag in state.TAGS.list_tags_for_resource(resource).get("Tags")
+        }
+        return lambda_adapted_tags
 
-    def _store_tags(self, function: Function, tags: dict[str, str]):
-        if len(tags) > LAMBDA_TAG_LIMIT_PER_RESOURCE:
+    def _store_tags(self, resource: TaggableResource, tags: dict[str, str]):
+        state = self.fetch_lambda_store_for_tagging(resource)
+        if len(state.TAGS.tags.get(resource, {}) | tags) > LAMBDA_TAG_LIMIT_PER_RESOURCE:
             raise InvalidParameterValueException(
                 "Number of tags exceeds resource tag limit.", Type="User"
             )
-        with function.lock:
-            function.tags = tags
-            # dirty hack for changed revision id, should reevaluate model to prevent this:
-            latest_version = function.versions["$LATEST"]
-            function.versions["$LATEST"] = dataclasses.replace(
-                latest_version, config=dataclasses.replace(latest_version.config)
-            )
 
-    def _update_tags(self, function: Function, tags: dict[str, str]):
-        with function.lock:
-            stored_tags = function.tags or {}
-            stored_tags |= tags
-            self._store_tags(function=function, tags=stored_tags)
+        tag_svc_adapted_tags = [{"Key": key, "Value": value} for key, value in tags.items()]
```

Comment:
We need to adapt the tags `TaggingService` <-> `LambdaProvider` when doing storage and retrieval of tags.

Reply:
I like the unified approach to tagging. Even though adapting in this way is a bit janky, I think the benefits of a single svc outweighs the costs of adapting to suit it.

In future, maybe we can extend the tag svc to better account for cases like this.